RecoTauTag/RecoTau/plugins/rerunMVAIsolationOnMiniAOD_Phase2.cc

```diff
@@ -0,0 +1,226 @@
+#include "FWCore/Framework/interface/EDAnalyzer.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+//#include "DataFormats/PatCandidates/interface/CompositeCandidate.h"
+#include "DataFormats/HepMCCandidate/interface/GenParticle.h"
+#include "DataFormats/PatCandidates/interface/Tau.h"
+#include "FWCore/ServiceRegistry/interface/Service.h"
+#include "CommonTools/UtilAlgos/interface/TFileService.h"
+
+#include "TH1D.h"
+#include "TH2D.h"
+#include "TGraphAsymmErrors.h"
+
+class rerunMVAIsolationOnMiniAOD_Phase2 : public edm::EDAnalyzer {
+public:
+  explicit rerunMVAIsolationOnMiniAOD_Phase2(const edm::ParameterSet&);
+private:
+   virtual void analyze(const edm::Event&, const edm::EventSetup&) override;
+   virtual void endJob() override;
+
+   edm::EDGetTokenT<pat::TauCollection> tauToken_;
+ 
+   TH1D *h_raw_bkg, *h_raw_sig;
+ 
+   TH1D *h_pt_denom_bkg, *h_pt_denom_sig;
+   TH1D *h_pt_num_bkg[7], *h_pt_num_sig[7];
+   TGraphAsymmErrors *eff[7], *fak[7];
+   
+   TH1D *h_inc_denom_bkg, *h_inc_num_bkg;
+   TH1D *h_inc_denom_sig, *h_inc_num_sig;
+   TGraphAsymmErrors *inceff, *incfak;
+
+   TH2D *h2_denom_bkg, *h2_denom_sig;
+   TH2D *h2_num_bkg[7], *h2_num_sig[7];
+   TH2D *eff2[7], *fak2[7];
+
+   bool haveGenJets;
+   edm::EDGetTokenT<std::vector<reco::GenJet>> genJetToken_;   
+   //bool haveGenVisTaus;
```

Comment:
I will remove this line.

Reply:
OK, great, it is what I wanted to suggest. You should however note that it builds genTaus also for tau->e/mu decays, which probably should be excluded here, see e.g. how it is used in PAT/miniAOD: https://github.com/cms-tau-pog/cmssw/blob/CMSSW_11_1_0_X_tau-pog_phase2-mvaiso/PhysicsTools/PatAlgos/python/producersLayer1/tauProducer_cff.py#L14-L15 and [L27-28](https://github.com/cms-tau-pog/cmssw/blob/CMSSW_11_1_0_X_tau-pog_phase2-mvaiso/PhysicsTools/PatAlgos/python/producersLayer1/tauProducer_cff.py#L27-L28).
You can also consider to use patTau.genJet() which receives pointer to matched tauVisJet (made as mentioned above) if found.

Reply:
Please correct me if I am mistaken:
Instead of explicitly producing the visible tau collection myself, I can simply ask if tau->genJet() is non-empty. Importantly, this genJet() only consists of hadronic tau decays.

That is seen in this list of tasks:
https://github.com/cms-tau-pog/cmssw/blob/CMSSW_11_1_0_X_tau-pog_phase2-mvaiso/PhysicsTools/PatAlgos/python/producersLayer1/tauProducer_cff.py#L26-L29

Reply:
Yes, it should work. The only issue is that matching to gen in patTaus is tighter than usually used (dR<0.1 & dPt/Pt<3 vs dR<0.2), see https://github.com/cms-tau-pog/cmssw/blob/CMSSW_11_1_0_X_tau-pog_phase2-mvaiso/PhysicsTools/PatAlgos/python/mcMatchLayer0/tauMatch_cfi.py#L21-L31. But, maybe for testing purposes it is not a real issue.